### PR TITLE
[SPARK-19695][SQL] Throw an exception if a `columnNameOfCorruptRecord` field violates requirements in json formats

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr comes from #16928 and fixed a json behaviour along with the CSV one. 

## How was this patch tested?
Added tests in `JsonSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
@HyukjinKwon @cloud-fan cloud you check this? thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 1]
```diff
@@ -1944,4 +1944,37 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
...
+        .option("columnNameOfCorruptRecord", columnNameOfCorruptRecord)
+        .schema(schema)
+        .json(corruptRecords)
+    }.getMessage
+    assert(errMsg.startsWith("The field for corrupt records must be string type and nullable"))
+
+    withTempPath { dir =>
+      val path = dir.getCanonicalPath
+      spark.createDataFrame(
+        corruptRecords.map(Row(_)), new StructType().add("value", StringType)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: `corruptRecords.toDF("value")`?

**Comment by HyukjinKwon** [comment]
Thanks for cc'ing me. I am okay if @cloud-fan is okay.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala [hunk 2]
```diff
@@ -102,6 +102,15 @@ class JsonFileFormat extends TextBasedFileFormat with DataSourceRegister {
       sparkSession.sessionState.conf.sessionLocalTimeZone,
       sparkSession.sessionState.conf.columnNameOfCorruptRecord)
 
+    // Check a field requirement for corrupt records here to throw an exception in a driver side
+    dataSchema.getFieldIndex(parsedOptions.columnNameOfCorruptRecord).map { corruptFieldIndex =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: `map` to `foreach`?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: And.. not a strong preference but maybe put this in `JacksonUtils` and rename it `JsonUtils` if @cloud-fan is okay? 

Maybe we could throws an exception as `IlligalArgumentException` in the first place and then capture the message with `AnalysisException` (as `JacksonUtils.verifySchema` is doing in `StructToJson`). This is not a strong opinion too.

**Comment by HyukjinKwon** [comment]
Thanks for cc'ing me. I am okay if @cloud-fan is okay.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: ideally the `columnNameOfCorruptRecord` stuff has nothing to do with parser. Parser just parses the record and report error if some records are bad, and the upper-level will handle the bad records and may put the bad record in a special string column.

I'm ok to keep this code snippet duplicated in 2 places.

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: okay!

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**